### PR TITLE
HashMap#updatedWith optimized to update in single lookup

### PR DESCRIPTION
**Pull request by joshlemer**
* Optimizes Hashmap#updatedWith method to perform the update transformation with only a single look-up, rather than the slower default implementation which requires two lookups (and as well, two key hashings)

### CODE REVIEW

**Comment by joshlemer** [comment]
Since this is so late, to avoid any risk, this change can go in in 2.13.1, if we're willing to stub the `HashMap` override of `updatedWith`

HashMap.scala:

```scala
override def updatedWith(...) = super.updatedWith(...)
```

**Comment by joshlemer** [comment]
## Benchmark

### Code

```scala

  case class Foo(i: Int, j: Int)

  @Param(Array("0", "1", "2", "3", "4", "5", "6", "7", "10", "100", "1000", "10000"))
  var size: Int = _

  var map: HashMap[Foo, String] = _
  var presentFoo: Foo = _

  val absentFoo = Foo(-1, -1)

  @Setup(Level.Trial)
  def setup(): Unit = {
    map = (1 to size).iterator.map(i => Foo(i, i) -> i.toString).to(HashMap)
    presentFoo = Foo(size, size)
  }



  @Benchmark
  def updateWithNew(bh: Blackhole): Unit = {
    // some to some (different)
    bh.consume(map.updatedWith(presentFoo) { case Some(value) => Some(value + "_") case None => None })

    // some to some (same)
    bh.consume(map.updatedWith(presentFoo)(value => value))
    // some to none
    bh.consume(map.updatedWith(absentFoo)(_ => None))

    // none to some
    bh.consume(map.updatedWith(absentFoo)(_ => Some("")))
    // none to none
    bh.consume(map.updatedWith(absentFoo)(_ => None))
  }
  @Benchmark
  def updateWithOld(bh: Blackhole): Unit = {
    // some to some (different)
    bh.consume(map.updatedWithOld(presentFoo) { case Some(value) => Some(value + "_") case None => None })

    // some to some (same)
    bh.consume(map.updatedWithOld(presentFoo)(value => value))
    // some to none
    bh.consume(map.updatedWithOld(absentFoo)(_ => None))

    // none to some
    bh.consume(map.updatedWithOld(absentFoo)(_ => Some("")))
    // none to none
    bh.consume(map.updatedWithOld(absentFoo)(_ => None))
  }
```

### Results
```
[info] Benchmark                        (size)  Mode  Cnt    Score    Error  Units
[info] HashMapBenchmark3.updateWithNew       0  avgt   10   90.813 ±  2.610  ns/op
[info] HashMapBenchmark3.updateWithNew       1  avgt   10  140.970 ± 35.197  ns/op
[info] HashMapBenchmark3.updateWithNew       2  avgt   10  138.513 ±  7.133  ns/op
[info] HashMapBenchmark3.updateWithNew       3  avgt   10  135.604 ±  1.775  ns/op
[info] HashMapBenchmark3.updateWithNew       4  avgt   10  136.506 ±  6.582  ns/op
[info] HashMapBenchmark3.updateWithNew       5  avgt   10  170.319 ± 15.192  ns/op
[info] HashMapBenchmark3.updateWithNew       6  avgt   10  187.179 ± 30.819  ns/op
[info] HashMapBenchmark3.updateWithNew       7  avgt   10  193.107 ± 43.554  ns/op
[info] HashMapBenchmark3.updateWithNew      10  avgt   10  148.794 ±  4.266  ns/op
[info] HashMapBenchmark3.updateWithNew     100  avgt   10  233.077 ±  2.313  ns/op
[info] HashMapBenchmark3.updateWithNew    1000  avgt   10  238.900 ± 10.006  ns/op
[info] HashMapBenchmark3.updateWithNew   10000  avgt   10  289.290 ±  2.622  ns/op
[info] HashMapBenchmark3.updateWithOld       0  avgt   10   93.556 ±  0.666  ns/op
[info] HashMapBenchmark3.updateWithOld       1  avgt   10  204.644 ±  6.581  ns/op
[info] HashMapBenchmark3.updateWithOld       2  avgt   10  227.476 ± 37.736  ns/op
[info] HashMapBenchmark3.updateWithOld       3  avgt   10  213.500 ±  5.886  ns/op
[info] HashMapBenchmark3.updateWithOld       4  avgt   10  199.281 ± 11.365  ns/op
[info] HashMapBenchmark3.updateWithOld       5  avgt   10  238.898 ±  2.281  ns/op
[info] HashMapBenchmark3.updateWithOld       6  avgt   10  233.961 ± 19.268  ns/op
[info] HashMapBenchmark3.updateWithOld       7  avgt   10  258.365 ± 11.852  ns/op
[info] HashMapBenchmark3.updateWithOld      10  avgt   10  219.270 ±  3.425  ns/op
[info] HashMapBenchmark3.updateWithOld     100  avgt   10  330.374 ±  4.633  ns/op
[info] HashMapBenchmark3.updateWithOld    1000  avgt   10  296.487 ±  8.697  ns/op
[info] HashMapBenchmark3.updateWithOld   10000  avgt   10  396.607 ± 13.583  ns/op
```

**Comment by joshlemer** [comment]
Actually discovered a bug in the implementation.. wip for now

**Comment by NthPortal** [comment]
I am definitely in favour of stubbing the override

File: src/library/scala/collection/immutable/HashMap.scala [hunk 1]
```diff
@@ -430,6 +433,9 @@ private[immutable] object MapNode {
 
   def empty[K, V]: BitmapIndexedMapNode[K, V] = EmptyMapNode.asInstanceOf[BitmapIndexedMapNode[K, V]]
 
+  def single[K, V](key: K, value: V, keyHash: Int, originalHash: Int): BitmapIndexedMapNode[K, V] =
+    new BitmapIndexedMapNode[K, V](Node.bitposFrom(Node.maskFrom(keyHash, 0)), 0, Array(key, value), Array(originalHash), 1, keyHash)
```
Comment (joshlemer) [inline src/library/scala/collection/immutable/HashMap.scala:436]: No need for a call to `Node.maskFrom(keyHash, 0))`, we can simplify to `Node.bitposFrom(keyHash)`

**Comment by lihaoyi** [comment]
@joshlemer should we pre-emptively override `updateWith` in `mutable.HashMap` so that too can be done in a single lookup?

**Comment by lihaoyi** [comment]
See https://github.com/scala/scala/pull/8073

**Comment by joshlemer** [comment]
@lihaoyi Yes I think it would be a great idea to preemptively-override the `mutable.HashMap`/`mutable.LinkedHashMap` versions.

**Comment by szeiger** [comment]
Rebased to get the nicer MiMa reports.

**Comment by szeiger** [comment]
The MiMa errors are in package-private classes. They should all be safe for whitelisting.

**Comment by retronym** [comment]
I'll review this tomorrow to try to see if its ready for 2.13.1

**Comment by lihaoyi** [comment]
whitelisted more things...

**Comment by SethTisue** [comment]
needs rebase

**Comment by szeiger** [comment]
@retronym ping

File: src/library/scala/collection/immutable/HashMap.scala [hunk 2]
```diff
@@ -1814,6 +1888,34 @@ private final class HashCollisionMapNode[K, +V ](
...
+        case Some(newValue) =>
+          if (newValue.asInstanceOf[AnyRef] eq v.asInstanceOf[AnyRef]) this
+          else new HashCollisionMapNode[K, V1](originalHash, hash, content.updated(index, (k, newValue)))
+        case None =>
+          if (content.length == 2) {
+            val remainingIndex = if (index == 0) 1 else 0
+            val (remainingKey, remainingValue) = content(remainingIndex)
+            MapNode.single(remainingKey, remainingValue, hash, originalHash)
+          } else {
+            new HashCollisionMapNode[K, V1](originalHash, hash, content.iterator.take(index).concat(content.iterator.drop(index + 1)).toVector)
```
Comment (retronym) [inline src/library/scala/collection/immutable/HashMap.scala:1906]: `content.patch(index, Nil, 1)` might be slightly clearer here.

**Comment by retronym** [comment]
@szeiger While this only needs a few minor cleanups, I've assigned it to 2.13.2 as I don't think we should hold up 2.13.1 for this. The performance gain is great, but `updatedWith` is a new part of the collections API so we won't be giving "free performance" to much existing code.

**Comment by joshlemer** [comment]
Sounds good to me

**Comment by szeiger** [comment]
@joshlemer Needs rebase

@retronym Any objections to merging this?

**Comment by SethTisue** [comment]
@joshlemer interested in returning to this? it still needs a rebase. we've marked it as draft for now

**Comment by dwijnand** [comment]
Closing for inactivity. Happy to reopen any time, by Josh or anyone else.

**Pull request closed without merging**